docs/userguide/torch_forecasting_models.md

```diff
@@ -461,6 +461,109 @@ model.fit(...)
+# any TorchMetric or val_loss can be used as the monitor
+torch_metrics = torchmetrics.regression.MeanAbsolutePercentageError()
+
+# MLflow setup
+## Run this command with environment activated: mlflow ui --port xxxx  (e.g. 5000, 5001, 5002)
+# Copy and paste url from command line to web browser
+import mlflow
+import torchmetrics
+from mlflow.data.pandas_dataset import PandasDataset
+
+mlflow.pytorch.autolog(log_every_n_epoch=1, log_every_n_step=None, 
+                        log_models=True, log_datasets=True, disable=False, 
+                        exclusive=False, disable_for_unsupported_versions=False, 
+                        silent=False, registered_model_name=None, extra_tags=None
+                        )
+
+import mlflow.pytorch
+from mlflow.client import MlflowClient
+
+model_name = "Darts"
+
+with mlflow.start_run(nested=True) as run:
+
+    dataset: PandasDataset = mlflow.data.from_pandas(series, source="AirPassengersDataset")
+
+    # Log the dataset to the MLflow Run. Specify the "training" context to indicate that the
+    # dataset is used for model training
+    mlflow.log_input(dataset, context="training")
+   
+    mlflow.log_param("model_type", "Darts_Pytorch_model")
+    mlflow.log_param("input_chunk_length", 24)
+    mlflow.log_param("output_chunk_length", 12)                   
+    mlflow.log_param("n_epochs", 500)
+    mlflow.log_param("model_name", 'NBEATS_MLflow')
+    mlflow.log_param("log_tensorboard", True)
+    mlflow.log_param("torch_metrics", "torchmetrics.regression.MeanAbsolutePercentageError()")
+    mlflow.log_param("nr_epochs_val_period", 1)
+    mlflow.log_param("pl_trainer_kwargs", "{callbacks: [loss_logger]}")    
+   
+   
+    from pytorch_lightning.callbacks import Callback
+
+    class LossLogger(Callback):
+        def __init__(self):
+            self.train_loss = []
+            self.val_loss = []
+
+        # will automatically be called at the end of each epoch
+        def on_train_epoch_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
+            self.train_loss.append(float(trainer.callback_metrics["train_loss"]))
+
+        def on_validation_epoch_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
+            self.val_loss.append(float(trainer.callback_metrics["val_loss"]))
+
+
+    loss_logger = LossLogger()    
+
+    # create the model
+    model = NBEATSModel(
+        input_chunk_length=24,
+        output_chunk_length=12,
+        n_epochs=500,
+        model_name='NBEATS_MLflow',
+        log_tensorboard=True,
+        torch_metrics=torch_metrics,
+        nr_epochs_val_period=1,
+        pl_trainer_kwargs={"callbacks": [loss_logger]})
+
+    # use validation dataset
+    model.fit(
+        series=train,
+        val_series=val,
+        )
+
+    # predit
+    forecast = model.predict(len(val))
+
+# Registering model
+model_uri = f"runs:/{run.info.run_id}/darts-NBEATS"
+mlflow.register_model(model_uri=model_uri, name=model_name)
```

Comment:
IMHO runs currently do not save models as artifacts becaus there is no call to `log_model()`. Therefore `model_uri` is not pointing to a valid model. Please correct me if I'm wrong, however, it did not work in my tests. If it works, can you pleas add an example how to load with ` loaded_model = mlflow.<flavor>.load_model(model_uri=model_uri)` ?

Reply:
As i understand it you can only (kind of) promote a already saved model to a registered model..

> Model
> An MLflow Model is created from an experiment or run that is logged with one of the model flavor’s mlflow.<model_flavor>.log_model() methods. **Once logged, this model can then be registered with the Model Registry.**
> 
> Registered Model
> An MLflow Model can be registered with the Model Registry. A registered model has a unique name, contains versions, aliases, tags, and other metadata.
https://mlflow.org/docs/latest/model-registry.html#concepts

Reply:
> As i understand it you can only (kind of) promote a already saved model to a registered model..
> 
> > Model
> > An MLflow Model is created from an experiment or run that is logged with one of the model flavor’s mlflow.<model_flavor>.log_model() methods. **Once logged, this model can then be registered with the Model Registry.**
> > Registered Model
> > An MLflow Model can be registered with the Model Registry. A registered model has a unique name, contains versions, aliases, tags, and other metadata.
> > https://mlflow.org/docs/latest/model-registry.html#concepts

Hello @turbotimon ,
My plans was to share how to do the first part of issue # 1618. As it allows you to track the experiments (so one don't get lost if you are doing more of a few iterations), select the best one, see metrics progress, etc.
I will have a look at what you are suggesting. However, the normal mlflow.python.loadmodel or mlflow.pytorch.loadmodel doesn't seems to work because the way Darts is wrapping torch.nn modules. All test I have done so far didn't work to save and load  the model. Only to record hyperparameters, other torchmetrics related information, dataset hash, etc. Thus, I propose to share this, and later we can add saving and loading models (but my last weekend research it is pointing out that the model.py files and common models may need to be rewritten to be compatible with MLFlow (I might be wrong).

Cheers!

Reply:
@cargecla1 Yes, that's a good idea to split these two things

If using the mlflow model registry for darts completely fails, you could also mentioning the workaround I proposed in the [issue](https://github.com/unit8co/darts/issues/1618#issuecomment-1753125634). Which was manually saving/loading the model as an artifact. Something like:
```python
dartsmodel.save("mymodel.pickle")
mlflow.log_artifact("mymodel.pickle")
# later, load/download artifact from mlflow and do e.g. a RNNModel.load("mymodel.pickle")
```

Let me know if i can help anything!

Reply:
> @cargecla1 Yes, that's a good idea to split these two things
> 
> If using the mlflow model registry for darts completely fails, you could also mentioning the workaround I proposed in the [issue](https://github.com/unit8co/darts/issues/1618#issuecomment-1753125634). Which was manually saving/loading the model as an artifact. Something like:
> 
> ```python
> dartsmodel.save("mymodel.pickle")
> mlflow.log_artifact("mymodel.pickle")
> # later, load artifact from mlflow and do e.g. a dartsmodel=RNNModel.load("mymodel.pickle")
> ```
> 
> Let me know if i can help anything!

Hello @turbotimon ,

Copy that, I forgot this comments from the issue you linked on your message.
This could work because you can save a darts model using pickle (and mlflow) and you can load models using pickle API.
Additional references to try:
https://mlflow.org/docs/latest/model-registry.html?highlight=pickle#id18
https://mlflow.org/docs/latest/model-registry.html?highlight=pickle#id19

Thank you!

Reply:
Hello @turbotimon 

After trying to save the Darts model using dartsmodel.save("mymodel.pickle") as suggested, I am getting the following error:

File ~\.conda\envs\venv_darts\lib\site-packages\torch\serialization.py:653, in _save(obj, zip_file, pickle_module, pickle_protocol)
    651 pickler = pickle_module.Pickler(data_buf, protocol=pickle_protocol)
    652 pickler.persistent_id = persistent_id
--> 653 pickler.dump(obj)
    654 data_value = data_buf.getvalue()
    655 zip_file.write_record('data.pkl', data_value, len(data_value))

AttributeError: Can't pickle local object 'LayerSummary._register_hook.<locals>.hook'

Would you know why this is happening?

Cheers

Reply:
> @cargecla1 Yes, that's a good idea to split these two things
> 
> If using the mlflow model registry for darts completely fails, you could also mentioning the workaround I proposed in the [issue](https://github.com/unit8co/darts/issues/1618#issuecomment-1753125634). Which was manually saving/loading the model as an artifact. Something like:
> 
> ```python
> dartsmodel.save("mymodel.pickle")
> mlflow.log_artifact("mymodel.pickle")
> # later, load artifact from mlflow and do e.g. a dartsmodel=RNNModel.load("mymodel.pickle")
> ```
> 
> Let me know if i can help anything!

Hello @turbotimon , @madtoinou , @dennisbader,

Is it possible to split this issue as discussed above? so we can share how to use, train, track, monitor and save the models using MLFlow? Just leaving loading the model for a later release?

Cheers

Reply:
> Hello @turbotimon
> 
> After trying to save the Darts model using dartsmodel.save("mymodel.pickle") as suggested, I am getting the following error:
> 
> File ~.conda\envs\venv_darts\lib\site-packages\torch\serialization.py:653, in _save(obj, zip_file, pickle_module, pickle_protocol) 651 pickler = pickle_module.Pickler(data_buf, protocol=pickle_protocol) 652 pickler.persistent_id = persistent_id --> 653 pickler.dump(obj) 654 data_value = data_buf.getvalue() 655 zip_file.write_record('data.pkl', data_value, len(data_value))
> 
> AttributeError: Can't pickle local object 'LayerSummary._register_hook..hook'
> 
> Would you know why this is happening?
> 
> Cheers

Hi @cla-ra3426, sorry i missed your question. No idea.. but must have to do somehting with darts itself and not mlflow i suppose

Reply:
> > @cargecla1 Yes, that's a good idea to split these two things
> > If using the mlflow model registry for darts completely fails, you could also mentioning the workaround I proposed in the [issue](https://github.com/unit8co/darts/issues/1618#issuecomment-1753125634). Which was manually saving/loading the model as an artifact. Something like:
> > ```python
> > dartsmodel.save("mymodel.pickle")
> > mlflow.log_artifact("mymodel.pickle")
> > # later, load artifact from mlflow and do e.g. a dartsmodel=RNNModel.load("mymodel.pickle")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Let me know if i can help anything!
> 
> Hello @turbotimon , @madtoinou , @dennisbader,
> 
> Is it possible to split this issue as discussed above? so we can share how to use, train, track, monitor and save the models using MLFlow? Just leaving loading the model for a later release?
> 
> Cheers

Hello gents (@turbotimon , @madtoinou , @dennisbader,)

Have you had time to consider my proposal above?  _"Is it possible to split this issue as discussed above? so we can share how to use, train, track, monitor and save the models using MLFlow? Just leaving loading the model for a later release?"_

The only solution I have right now it is to train the model with MLFlow to track and monitor the model and retrain it with pure Darts to be able to save it and load to predict, Darts saving method "fails" when run inside a MLFlow run, and MLFlow log and save methods don't want to work with Darts either.

Thank you in advice for your consideration!

Cheers!

Reply:
Hi @cargecla1,

Sorry for the delay, loading model is kind-of part of linked issue so it would probably be better to also include it in this PR. But if it too troublesome, we can treat it separately. 

@cla-ra3426,

The error you're getting when trying to pickle the model is probably due to the (pytorch-lightning) callbacks. Can you try removing them prior to exporting the model?

I will test this example more thoroughly when I have more time, try to see if I can come up with a solution for the loading.

Reply:
Hello @madtoinou 

I will try this and see how it goes.

Thanks for this suggestion and for considering splitting the problem into two if above doesn't work.

Reply:
Hello @madtoinou ,

I took your suggestion on board as per above, but that didn't fixed the issue with loading and predicting steps.

Refer to commit https://github.com/unit8co/darts/pull/2092/commits/f7982440b876ff6d07761a2ed6ac8717fe93a1c0